cmd/clusterctl/client/repository/components.go

```diff
@@ -493,78 +484,6 @@ func fixCertificate(o unstructured.Unstructured, originalNamespace, targetNamesp
 	return o, nil
 }
 
-// fixRBAC ensures all the ClusterRole and ClusterRoleBinding have the name prefixed with the namespace name and that
-// all the clusterRole/clusterRoleBinding namespaced subjects refers to targetNamespace.
-func fixRBAC(objs []unstructured.Unstructured, targetNamespace string) ([]unstructured.Unstructured, error) {
-	renamedClusterRoles := map[string]string{}
-	for _, o := range objs {
-		// if the object has Kind ClusterRole
-		if o.GetKind() == clusterRoleKind {
-			// assign a namespaced name
-			currentName := o.GetName()
-			newName := fmt.Sprintf("%s-%s", targetNamespace, currentName)
-			o.SetName(newName)
-
-			renamedClusterRoles[currentName] = newName
-		}
-	}
-
-	for i := range objs {
-		o := objs[i]
-		switch o.GetKind() {
-		case clusterRoleBindingKind: // if the object has Kind ClusterRoleBinding
-			// Convert Unstructured into a typed object
-			b := &rbacv1.ClusterRoleBinding{}
-			if err := scheme.Scheme.Convert(&o, b, nil); err != nil {
-				return nil, err
-			}
-
-			// assign a namespaced name
-			b.Name = fmt.Sprintf("%s-%s", targetNamespace, b.Name)
-
-			// ensure that namespaced subjects refers to targetNamespace
-			for k := range b.Subjects {
-				if b.Subjects[k].Namespace != "" {
-					b.Subjects[k].Namespace = targetNamespace
```

Comment:
@fabriziopandini Should there also be a cleanup here?
https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/client/cluster/components.go#L181-L193

For example `!strings.HasPrefix(obj.GetName(), instanceNamespacePrefix)`

Reply:
good catch
Yes I think we should drop !strings.HasPrefix(obj.GetName(), instanceNamespacePrefix) and fix the comment

the correct way to test this is
- install provider with old clusterctl version
- upgrade provider using clusterctl from main, check all the old RBAC objects are deleted, the new one added (it could help have a breakpoint in the middle of upgrade)
- delete provider using provider, check  all the old RBAC objects are deleted

Reply:
There are some test cases that will fail if `!strings.HasPrefix(obj.GetName(), instanceNamespacePrefix)` is deleted. For example:

```
--- FAIL: Test_providerComponents_Delete (0.02s)
    --- FAIL: Test_providerComponents_Delete/Delete_provider_while_preserving_Namespace_and_CRDs_and_providerInventory (0.02s)
        components_test.go:298: /ns1-cluster-role not deleted, expect deleted
    --- FAIL: Test_providerComponents_Delete/Delete_provider_and_provider_namespace,_while_preserving_CRDs_and_providerInventory (0.00s)
        components_test.go:298: /ns1-cluster-role not deleted, expect deleted
    --- FAIL: Test_providerComponents_Delete/Delete_provider_and_provider_CRDs,_while_preserving_the_provider_namespace_and_the_providerInventory (0.00s)
        components_test.go:298: /ns1-cluster-role not deleted, expect deleted
    --- FAIL: Test_providerComponents_Delete/Delete_providerInventory_and_provider_while_preserving_provider_CRDs_and_provider_namespace (0.00s)
        components_test.go:298: /ns1-cluster-role not deleted, expect deleted
    --- FAIL: Test_providerComponents_Delete/Delete_provider,_provider_namespace_and_provider_CRDs_and_the_providerInventory (0.00s)
        components_test.go:298: /ns1-cluster-role not deleted, expect deleted

```

Link to test cases: https://github.com/kubernetes-sigs/cluster-api/blob/6b042456383aa18ba61fc5de5c41901dd7267c60/cmd/clusterctl/client/cluster/components_test.go#L161

So there are a number of existing use cases where CAPI expects cluster roles to be deleted. Maybe this check should remain for now? What we want to stop is creating new prefixed cluster objects. But deleting them should maybe work as before for a while longer for the users that might actually have this?

Reply:
I need some time to dig into this, but ideally, we should be able to
- delete ClusterRole is named `ns1-cluster-role` (with the namespace as a prefix, installed with the older version of clusterctl)
- delete ClusterRole is named `cluster-role` (without the namespace as a prefix, installed with the version of clusterctl being currently developed)

If we need to preserve code like !strings.HasPrefix(obj.GetName(), instanceNamespacePrefix) to handle the first case let's make it sure we document why we are doing this, and possibly also when we can drop this case - that will be when the min version of CAPI we upgrade from is v1.3 -

Hope this helps to unblock, otherwise I will try to get back to this PR before EOW

Reply:
Thank you for the pointers. From my side there is no rush so please take your time.

I will leave the delete logic as is, but with a comment there references this pr and the issue. This will allow the deletion of resources that have been created with fixRBAC but at the same time removing the functionality to create more of those resources.

What remains then is also to create a new issue to clean this up when the min version to upgrade from is v1.3